### PR TITLE
Refactor `MetastoreError`

### PR DESCRIPTION
**Pull request by guilload**
### Description
- Move `Metastore{Error, Result}` to `quickwit-proto` crate
- Refactor `MetastoreError`

### How was this PR tested?
`make test-all`

### CODE REVIEW

File: quickwit/quickwit-index-management/src/garbage_collection.rs [hunk 1]
```diff
@@ -652,9 +652,9 @@ mod tests {
 
         let mut mock_metastore = MockMetastore::new();
         mock_metastore.expect_delete_splits().return_once(|_, _| {
-            Err(MetastoreError::IndexDoesNotExist {
+            Err(MetastoreError::NotFound(EntityKind::Index {
```
Comment (fmassot) [inline quickwit/quickwit-index-management/src/garbage_collection.rs:655]: niceee

File: quickwit/quickwit-proto/src/metastore/mod.rs [hunk 2]
```diff
@@ -17,7 +17,179 @@
...
+    },
+    /// A split.
+    Split {
+        /// Split ID.
+        split_id: SplitId,
+    },
+    /// A set of splits.
+    Splits {
+        /// Split IDs.
+        split_ids: Vec<String>,
```
Comment (fmassot) [inline quickwit/quickwit-proto/src/metastore/mod.rs:71]:
```suggestion
        split_ids: Vec<SplitID>,
```

File: quickwit/quickwit-proto/src/metastore/mod.rs [hunk 3]
```diff
@@ -17,7 +17,179 @@
...
+    CheckpointDelta {
+        /// Index ID.
+        index_id: IndexId,
+        /// Source Id.
+        source_id: SourceId,
+    },
+    /// An index.
+    Index {
+        /// Index ID.
+        index_id: IndexId,
```
Comment (fmassot) [inline quickwit/quickwit-proto/src/metastore/mod.rs:46]: An index also has a UID. I guess in some cases we would want to refer to the UID.

File: quickwit/quickwit-proto/src/metastore/mod.rs [hunk 4]
```diff
@@ -17,7 +17,179 @@
...
+        }
+    }
+}
+
+#[derive(Debug, Clone, thiserror::Error, Eq, PartialEq, Serialize, Deserialize)]
+pub enum MetastoreError {
+    #[error("Database error: `{message}`")]
+    Db { message: String },
+
+    #[error("Connection error: `{message}`")]
```
Comment (fmassot) [inline quickwit/quickwit-proto/src/metastore/mod.rs:116]: dot or no dot.

```suggestion
    #[error("Connection error: `{message}`.")]
```

File: quickwit/quickwit-proto/src/metastore/mod.rs [hunk 5]
```diff
@@ -17,7 +17,179 @@
...
+    Connection { message: String },
+
+    #[error("Precondition failed: {message}")]
```
Comment (fmassot) [inline quickwit/quickwit-proto/src/metastore/mod.rs:119]:
```suggestion
    #[error("Precondition failed on {entity}: {message}")]
```

**Review by fmassot — APPROVED** [review]

File: quickwit/quickwit-proto/src/metastore/mod.rs (hunk 3 above)
Comment (guilload) [reply quickwit/quickwit-proto/src/metastore/mod.rs:46]: True. We kept displaying index IDs to the users when we introduced index UIDs, so I followed that convention.

File: quickwit/quickwit-proto/src/metastore/mod.rs [hunk 6]
```diff
@@ -17,7 +17,179 @@
...
+            EntityKind::Source { source_id } => write!(f, "Source `{source_id}`"),
+            EntityKind::Split { split_id } => write!(f, "Split `{split_id}`"),
+            EntityKind::Splits { split_ids } => write!(f, "Splits `{}`", split_ids.join(", ")),
+        }
+    }
+}
+
+#[derive(Debug, Clone, thiserror::Error, Eq, PartialEq, Serialize, Deserialize)]
+pub enum MetastoreError {
+    #[error("Database error: `{message}`")]
```
[the remark this answers is not in the document]
Comment (guilload) [reply quickwit/quickwit-proto/src/metastore/mod.rs:113]: I guess dot xD

**Pull request merged**